src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4302,6 +4302,48 @@ void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, XMMRegister
 }
 #endif
 
+void C2_MacroAssembler::vector_signum_evex(int opcode, XMMRegister dst, XMMRegister src, XMMRegister zero, XMMRegister one,
+                                           XMMRegister xtmp1, KRegister ktmp1, KRegister ktmp2, int vec_enc) {
+  if (opcode == Op_SignumVD) {
+    evfpclasspd(ktmp1, src, 0x7, vec_enc);
+    evpmovq2m(ktmp2, src, vec_enc);
+    evsubpd(xtmp1, k0, zero, one, true, vec_enc);
+    evmovdquq(dst, k0, one, true, vec_enc);
+    evmovdquq(dst, ktmp2, xtmp1, true, vec_enc);
+    evblendmpd(dst, ktmp1, dst, src, true, vec_enc);
+  } else {
+    assert(opcode == Op_SignumVF, "");
+    evfpclassps(ktmp1, src, 0x7, vec_enc);
+    evpmovd2m(ktmp2, src, vec_enc);
+    evsubps(xtmp1, k0, zero, one, true, vec_enc);
+    evmovdqul(dst, k0, one, true, vec_enc);
+    evmovdqul(dst, ktmp2, xtmp1, true, vec_enc);
+    evblendmps(dst, ktmp1, dst, src, true, vec_enc);
+  }
+}
+
+void C2_MacroAssembler::vector_signum_avx(int opcode, XMMRegister dst, XMMRegister src, XMMRegister zero, XMMRegister one,
+                                          XMMRegister xtmp1, XMMRegister xtmp2, XMMRegister xtmp3, int vec_enc) {
+  if (opcode == Op_SignumVD) {
+    vcmppd(xtmp1, src, src, Assembler::UNORD_Q, vec_enc);
+    vcmppd(xtmp2, src, zero, Assembler::EQ_OQ, vec_enc);
+    vpor(xtmp1, xtmp1, xtmp2, vec_enc);
+    vcmppd(xtmp2, zero, src, Assembler::LT_OS, vec_enc);
+    vsubpd(xtmp3, zero, one, vec_enc);
+    vblendvpd(dst, xtmp3, one, xtmp2, vec_enc);
+    vblendvpd(dst, dst, src, xtmp1, vec_enc);
```

Comment:
The same applies here, suggestion:

    vsubpd(dst, zero, one, vec_enc);
    vcmppd(xtmp1, src, zero, Assembler::LT_OQ, vec_enc);
    vblendvpd(dst, xtmp1, one, dst, vec_enc);
    vcmppd(xtmp1, src, zero, Assembler::EQ_UQ, vec_enc);
    vblendvpd(dst, xtmp1, dst, zero, vec_enc);

Thanks.

Reply:
Same responses as above.

Reply:
I don't fully understand what you mean by "same as above". The logic here is using `src` to blend `one` and `minus_one` is that every positive will get `one` and every negative will get `minus_one`, the others are `zero` and `Nan` will make the unordered comparison with zero return true, thus the final blend will result in the correct output.
Thanks.

Reply:
Good, I recall using this in one of my previous patch 
http://hg.openjdk.java.net/jdk/jdk/rev/ff399127078a#l5.245